src/Input/Tokenizer.php

```diff
@@ -0,0 +1,464 @@
+<?php
+
+/*
+ * This file is part of the PHP-CLI package.
+ *
+ * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
+ *     <https://github.com/adhocore>
+ *
+ * Licensed under MIT license.
+ */
+
+namespace Ahc\Cli\Input;
+
+use Ahc\Cli\Input\Token;
+use \Iterator;
+
+use function \explode;
+use function \array_map;
+use function \array_push;
+use function \ltrim;
+use function \rtrim;
+use function \preg_match;
+use function \preg_quote;
+use function \sprintf;
+use function \str_split;
+use function \strlen;
+
+/**
+ * Tokenizer.
+ * A tokenizer is a class that takes an array of arguments and
+ * converts them into pre-defined tokens.
+ *
+ * @author  Shlomo Hassid <shlomohassid@gmail.com>
+ * @license MIT
+ *
+ * @link    https://github.com/adhocore/cli
+ */
+class Tokenizer implements Iterator
+{
+    
+    /** @var Token[]  */
+    private array $tokens = [];
+
+    private int $index = 0;
+
+    /**
+     * @param array $args
+     */
+    public function __construct(array $args) 
+    {
+        // Flags:
+        $variadic = false;
+        $literal  = false;
+
+        // Process args:
+        foreach ($args as $arg) {
+
+            // Tokenize this arg:
+            $tokens = [];
+            if (
+                // Not a literal token:
+                !$literal || 
+                // Or a literal token with variadic close:
+                // Which is a special case:
+                ($variadic && ($arg[0] ?? '') === Token::TOKEN_VARIADIC_C)
+            ) {
+                
+                $tokens = $this->tokenize($arg);
+                $literal = false;
+
+            } else {
+                // Literal token treat all as constant:
+                $tokens[] = new Token(Token::TYPE_CONSTANT, $arg);
+            }
+
+            // Process detected token/s:
+            foreach ($tokens as $token) {
+            
+                switch ($token->type()) {
+            
+                    case Token::TYPE_VARIADIC:
+                        $variadic = $token->isVariadic("open");
+                        if ($variadic) {
+                            $this->tokens[] = $token;
+                        }
+                        break;
+
+                    case Token::TYPE_LITERAL:
+                        // Add all remaining tokens as nested of a literal token:
+                        $literal = true;
+                        break;
+
+                    case Token::TYPE_SHORT:
+                    case Token::TYPE_LONG:
+                    case Token::TYPE_CONSTANT:
+                        if ($variadic) {
+                            $this->tokens[count($this->tokens) - 1]->addNested($token);
```

Comment:
the nesting goes quite deep: `class -> func -> foreach -> foreach -> switch -> case -> if` 
probably need to optimize, specially loop inside loop

Reply:
What is wrong with that? It's a lot cheaper than function/method calls...
1 scope block + 4 code blocks are not considered as deep I think.